File: brownian_motion_sim/brownian_motion.cpp
// Brownian motion (Wiener process) simulator.
//
// Simulates N independent particles undergoing d-dimensional Brownian
// motion via the discretized SDE:  x(t+dt) = x(t) + sqrt(2*D*dt) * N(0,1)
//
// Build:   g++ -O2 -std=c++17 -o brownian_motion brownian_motion.cpp
// Run:     ./brownian_motion --particles 200 --steps 1000 --dt 0.01 --D 1.0 --dim 2 --out trajectories.csv

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cmath>
#include <random>
#include <vector>
#include <string>
#include <fstream>
#include <iostream>

struct Config {
    int particles = 100;
    int steps = 1000;
    double dt = 0.01;
    double D = 1.0;      // diffusion coefficient
    int dim = 2;         // spatial dimension (1, 2, or 3)
    unsigned seed = std::random_device{}();
    std::string out = "trajectories.csv";
};

static void printUsage(const char* prog) {
    std::cout <<
        "Usage: " << prog << " [options]\n"
        "  --particles N   number of particles (default 100)\n"
        "  --steps N       number of time steps (default 1000)\n"
        "  --dt X          time step size (default 0.01)\n"
        "  --D X           diffusion coefficient (default 1.0)\n"
        "  --dim N         spatial dimension: 1, 2, or 3 (default 2)\n"
        "  --seed N        RNG seed (default random)\n"
        "  --out FILE      output CSV path (default trajectories.csv)\n"
        "  --help          show this message\n";
}

static bool parseArgs(int argc, char** argv, Config& cfg) {
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        auto next = [&](const char* name) -> std::string {
            if (i + 1 >= argc) {
                std::cerr << "Missing value for " << name << "\n";
                std::exit(1);
            }
            return argv[++i];
        };
        if (arg == "--particles") cfg.particles = std::stoi(next("--particles"));
        else if (arg == "--steps") cfg.steps = std::stoi(next("--steps"));
        else if (arg == "--dt") cfg.dt = std::stod(next("--dt"));
        else if (arg == "--D") cfg.D = std::stod(next("--D"));
        else if (arg == "--dim") cfg.dim = std::stoi(next("--dim"));
        else if (arg == "--seed") cfg.seed = static_cast<unsigned>(std::stoul(next("--seed")));
        else if (arg == "--out") cfg.out = next("--out");
        else if (arg == "--help" || arg == "-h") { printUsage(argv[0]); std::exit(0); }
        else { std::cerr << "Unknown argument: " << arg << "\n"; printUsage(argv[0]); return false; }
    }
    if (cfg.dim < 1 || cfg.dim > 3) {
        std::cerr << "--dim must be 1, 2, or 3\n";
        return false;
    }
    if (cfg.particles <= 0 || cfg.steps <= 0 || cfg.dt <= 0.0 || cfg.D < 0.0) {
        std::cerr << "particles/steps/dt must be positive, D must be non-negative\n";
        return false;
    }
    return true;
}

int main(int argc, char** argv) {
    Config cfg;
    if (!parseArgs(argc, argv, cfg)) return 1;

    std::mt19937_64 rng(cfg.seed);
    std::normal_distribution<double> normal(0.0, 1.0);
    const double stepStd = std::sqrt(2.0 * cfg.D * cfg.dt);

    // positions[p][k] = current coordinate k of particle p
    std::vector<std::vector<double>> positions(cfg.particles, std::vector<double>(cfg.dim, 0.0));

    std::ofstream csv(cfg.out);
    if (!csv) {
        std::cerr << "Failed to open output file: " << cfg.out << "\n";
        return 1;
    }
    csv << "step,time,particle";
    static const char* axisNames[3] = {"x", "y", "z"};
    for (int k = 0; k < cfg.dim; ++k) csv << "," << axisNames[k];
    csv << "\n";

    auto writeRow = [&](int step, double t, int p) {
        csv << step << "," << t << "," << p;
        for (int k = 0; k < cfg.dim; ++k) csv << "," << positions[p][k];
        csv << "\n";
    };

    for (int p = 0; p < cfg.particles; ++p) writeRow(0, 0.0, p);

    for (int step = 1; step <= cfg.steps; ++step) {
        double t = step * cfg.dt;
        for (int p = 0; p < cfg.particles; ++p) {
            for (int k = 0; k < cfg.dim; ++k) {
                positions[p][k] += stepStd * normal(rng);
            }
            writeRow(step, t, p);
        }
    }
    csv.close();

    // Sanity check: estimate D from the mean squared displacement at the
    // final time and compare against the configured value.
    double msd = 0.0;
    for (int p = 0; p < cfg.particles; ++p) {
        double sq = 0.0;
        for (int k = 0; k < cfg.dim; ++k) sq += positions[p][k] * positions[p][k];
        msd += sq;
    }
    msd /= cfg.particles;
    double finalT = cfg.steps * cfg.dt;
    double D_est = msd / (2.0 * cfg.dim * finalT);

    std::cout << "Simulated " << cfg.particles << " particles, " << cfg.steps
              << " steps, dim=" << cfg.dim << ", dt=" << cfg.dt
              << ", D=" << cfg.D << ", seed=" << cfg.seed << "\n";
    std::cout << "Final MSD = " << msd << "  ->  estimated D = " << D_est
              << " (configured D = " << cfg.D << ")\n";
    std::cout << "Wrote trajectories to " << cfg.out << "\n";

    return 0;
}

File: brownian_motion_sim/julia/brownian_motion.jl
#!/usr/bin/env julia
# Brownian motion (Wiener process) simulator.
#
# Simulates N independent particles undergoing d-dimensional Brownian
# motion via the discretized SDE:  x(t+dt) = x(t) + sqrt(2*D*dt) * N(0,1)
#
# Run:   julia brownian_motion.jl --particles 200 --steps 1000 --dt 0.01 --D 1.0 --dim 2 --out trajectories.csv
#
# Port of brownian_motion.cpp. The RNG stream will not match the C++ binary
# bit-for-bit -- Julia's default RNG (Xoshiro256++) and normal-variate
# algorithm differ from libstdc++'s mt19937_64 + std::normal_distribution --
# but both draw from the same N(0,1) distribution, so the simulated process
# and its statistics (e.g. the recovered diffusion coefficient) match.
#
# Particle indices in the output CSV are 0-indexed to match the original
# tool's format, even though Julia arrays are 1-indexed internally.

using Random
using Printf

mutable struct Config
    particles::Int
    steps::Int
    dt::Float64
    D::Float64          # diffusion coefficient
    dim::Int            # spatial dimension (1, 2, or 3)
    seed::UInt64
    out::String
end

Config() = Config(100, 1000, 0.01, 1.0, 2, rand(RandomDevice(), UInt64), "trajectories.csv")

function print_usage(prog::AbstractString)
    println("""
    Usage: $prog [options]
      --particles N   number of particles (default 100)
      --steps N       number of time steps (default 1000)
      --dt X          time step size (default 0.01)
      --D X           diffusion coefficient (default 1.0)
      --dim N         spatial dimension: 1, 2, or 3 (default 2)
      --seed N        RNG seed (default random)
      --out FILE      output CSV path (default trajectories.csv)
      --help          show this message""")
end

function parse_config(args::Vector{String})
    cfg = Config()
    prog = PROGRAM_FILE == "" ? "brownian_motion.jl" : basename(PROGRAM_FILE)

    i = 1
    next_val = (name) -> begin
        if i + 1 > length(args)
            println(stderr, "Missing value for $name")
            exit(1)
        end
        i += 1
        args[i]
    end

    while i <= length(args)
        arg = args[i]
        if arg == "--particles"
            cfg.particles = parse(Int, next_val("--particles"))
        elseif arg == "--steps"
            cfg.steps = parse(Int, next_val("--steps"))
        elseif arg == "--dt"
            cfg.dt = parse(Float64, next_val("--dt"))
        elseif arg == "--D"
            cfg.D = parse(Float64, next_val("--D"))
        elseif arg == "--dim"
            cfg.dim = parse(Int, next_val("--dim"))
        elseif arg == "--seed"
            cfg.seed = parse(UInt64, next_val("--seed"))
        elseif arg == "--out"
            cfg.out = next_val("--out")
        elseif arg == "--help" || arg == "-h"
            print_usage(prog)
            exit(0)
        else
            println(stderr, "Unknown argument: $arg")
            print_usage(prog)
            exit(1)
        end
        i += 1
    end

    if cfg.dim < 1 || cfg.dim > 3
        println(stderr, "--dim must be 1, 2, or 3")
        exit(1)
    end
    if cfg.particles <= 0 || cfg.steps <= 0 || cfg.dt <= 0.0 || cfg.D < 0.0
        println(stderr, "particles/steps/dt must be positive, D must be non-negative")
        exit(1)
    end
    return cfg
end

const AXIS_NAMES = ("x", "y", "z")

function simulate(cfg::Config)
    rng = Xoshiro(cfg.seed)
    step_std = sqrt(2.0 * cfg.D * cfg.dt)

    # positions[k, p] = current coordinate k of particle p
    positions = zeros(Float64, cfg.dim, cfg.particles)

    io = open(cfg.out, "w")
    print(io, "step,time,particle")
    for k in 1:cfg.dim
        print(io, ",", AXIS_NAMES[k])
    end
    print(io, "\n")

    function write_row(step, t, p)
        print(io, step, ",", t, ",", p - 1)
        for k in 1:cfg.dim
            print(io, ",", positions[k, p])
        end
        print(io, "\n")
    end

    for p in 1:cfg.particles
        write_row(0, 0.0, p)
    end

    for step in 1:cfg.steps
        t = step * cfg.dt
        for p in 1:cfg.particles
            for k in 1:cfg.dim
                positions[k, p] += step_std * randn(rng)
            end
            write_row(step, t, p)
        end
    end
    close(io)

    # Sanity check: estimate D from the mean squared displacement at the
    # final time and compare against the configured value.
    msd = 0.0
    for p in 1:cfg.particles
        sq = 0.0
        for k in 1:cfg.dim
            sq += positions[k, p]^2
        end
        msd += sq
    end
    msd /= cfg.particles
    final_t = cfg.steps * cfg.dt
    D_est = msd / (2.0 * cfg.dim * final_t)

    @printf("Simulated %d particles, %d steps, dim=%d, dt=%g, D=%g, seed=%d\n",
            cfg.particles, cfg.steps, cfg.dim, cfg.dt, cfg.D, cfg.seed)
    @printf("Final MSD = %g  ->  estimated D = %g (configured D = %g)\n",
            msd, D_est, cfg.D)
    println("Wrote trajectories to ", cfg.out)
end

function main()
    cfg = parse_config(ARGS)
    simulate(cfg)
end

main()
